File: Sem3/AlgAdat2/Code/avl.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <math.h>
#include "avl.h"

typedef struct Node {
	int key;
	int b;
	struct Node* left;
	struct Node* right;
} Node;

Node* new_node(int key) {
	Node* node = (Node*) malloc(sizeof(Node));

	if (node == NULL) {
		printf("FAILED TO CREATE NODE\n");
		exit(1);		
	}

	node->key = key;
	node->b = 0;
	node->left = NULL;
	node->right = NULL;

	return node;
}

void AVLinsert(Node** t, int k, bool* d) {
	if (*t == NULL) {
		(*t) = new_node(k);
		*d = true;
	}
	else if (k < (*t)->key) {
		AVLinsert(&(*t)->left, k, d);

		if (*d) {
			leftSubTreeGrown(t, d);
		}
	}
	else if (k > (*t)->key) {
		AVLinsert(&(*t)->right, k, d);

		if (*d) {
			rightSubTreeGrown(t, d);
		}
	}
	else {
		*d = false;
	}
}

void leftSubTreeGrown(Node** t, bool* d) {
	if ((*t)->b == -1) {
		Node* l = (*t)->left;

		if (l->b == -1) {
			balanceMMm(t, &l);
		}
		else {
			balanceMMp(t, &l);
		}

		*d = false;
	}
	else {
		(*t)->b -= 1;
		*d = ((*t)->b < 0);
	}
}

void rightSubTreeGrown(Node** t, bool* d) {
	if ((*t)->b == 1) {
		Node* r = (*t)->right;

		if (r->b == 1) {
			balancePPp(t, &r);
		}
		else {
			balancePPm(t, &r);
		}

		*d = false;
	}
	else {
		(*t)->b += 1;
		*d = ((*t)->b > 0);
	}
}

void balanceMMm(Node** t, Node** l) {
	(*t)->left = (*l)->right;
	(*l)->right = *t;
	(*t)->b = 0;
	(*l)->b = (*t)->b;
	*t = *l;
}

void balanceMMp(Node** t, Node** l) {
	Node* r = (*l)->right;
	(*t)->left = r->right;
	(*l)->right = r->left;
	r->left = (*l);
	r->right = (*t);
	(*l)->b = -floor((r->b + 1) / 2);
	(*t)->b = floor((1 - r->b) / 2);
	r->b = 0;
	*t = r;
}

void balancePPp(Node** t, Node** r) {
	(*t)->right = (*r)->left;
	(*r)->left = *t;
	(*t)->b = 0;
	(*r)->b = (*t)->b;
	*t = *r;
}

void balancePPm(Node** t, Node** r) {
	Node* l = (*r)->left;
	(*t)->right = l->left;
	(*r)->left = l->right;
	l->left = (*t);
	l->right = (*r);
	(*t)->b = -(floor((l->b + 1) / 2));
	(*r)->b = floor((1 - l->b) / 2);
	l->b = 0;
	*t = l;
}

void inorder_print(Node* t, int n) {
	if (t != NULL) {
		printf("(");

		if (t->left != NULL)
			inorder_print(t->left, n + 1);

		printf(" [%d] %d ", n, t->key);

		if (t->right != NULL)
			inorder_print(t->right, n + 1);

		printf(")");
	}
}

void print(Node* t) {
	inorder_print(t, 0);
	printf("\n");
}

void delete_nodes(Node* t) {
	if (t == NULL) {
		return;
	}

	delete_nodes(t->left);
	delete_nodes(t->right);
	free(t);
}

File: Sem3/AlgAdat2/Vizsga/AVL/avl/avl.go
package avl

import (
	"fmt"
	"math"
)

type Ordered interface {
    int | int8 | int16 | int32 | int64 | uint | uint8 | uint16 | uint32 | uint64 | float32 | float64 | string
}

type Node[T Ordered] struct {
    key         T
    b           int
    Left, Right *Node[T]
}

func NewNode[T Ordered]() *Node[T] {
    return &Node[T] {
        b: 0,
        Left: nil,
        Right: nil,
    }
}

func NewNodeKey[T Ordered](k T) *Node[T] {
    return &Node[T] {
        key: k,
        b: 0,
        Left: nil,
        Right: nil,
    }
}

func (node Node[T]) String() string {
    s := "["

    if node.Left != nil {
        s += node.Left.String()       
    }

    s += fmt.Sprintf(" %v(%v) ", node.key, node.b)

    if node.Right != nil {
        s += node.Right.String()
    }
    
    s += "]"
    return s
}

func AVLInsert[T Ordered](t **Node[T], k T, d *bool) {
    if *t == nil {
        *t = NewNodeKey(k)
        *d = true
    } else {
        if k < (*t).key {
            AVLInsert(&(*t).Left, k, d)
            if *d {
                leftSubTreeGrown(t, d)
            }
        } else if k > (*t).key {
            AVLInsert(&(*t).Right, k, d)
            if *d {
                rightSubTreeGrown(t, d)
            }
        } else {
            *d = false
        }
    }
}

func leftSubTreeGrown[T Ordered](t **Node[T], d *bool) {
    if (*t).b == -1 {
        l := (*t).Left
        if l.b == -1 {
            balanceMMm(t, &l)
        } else {
            balanceMMp(t, &l)
        }
        *d = false
    } else {
        (*t).b = (*t).b - 1
        *d = (*t).b < 0
    }
}

func rightSubTreeGrown[T Ordered](t **Node[T], d *bool) {
    if (*t).b == 1 {
        r := (*t).Right
        if r.b == 1 {
            balancePPp(t, &r)
        } else {
            balancePPm(t, &r)
        }
        *d = false
    } else {
        (*t).b = (*t).b + 1
        *d = (*t).b > 0
    }
}

func balancePPp[T Ordered](t, r **Node[T]) {
    (*t).Right = (*r).Left
    (*r).Left = *t
    (*t).b, (*r).b = 0, 0
    *t = *r
}

func balanceMMm[T Ordered](t, l **Node[T]) {
    (*t).Left = (*l).Right
    (*l).Right = *t
    (*t).b, (*l).b = 0, 0
    *t = *l
}

func balancePPm[T Ordered](t, r **Node[T]) {
    l := (*r).Left
    (*t).Right = l.Left
    (*r).Left = l.Right
    l.Left = *t
    l.Right = *r
    (*t).b = -int(math.Floor(float64(l.b + 1) / 2))   
    (*r).b = int(math.Floor(float64(1 - l.b) / 2))
    l.b = 0
    *t = l
}

func balanceMMp[T Ordered](t, l **Node[T]) {
    r := (*l).Right
    (*l).Right = r.Left
    (*t).Left = r.Right
    r.Left = *l
    r.Right = *t
    (*l).b = -int(math.Floor(float64(r.b + 1) / 2))
    (*t).b = int(math.Floor(float64(1 - r.b) / 2))
    r.b = 0
    *t = r
}
